File: src/main/java/controller/ModelController.java
package controller;

import database.Database;
import model.Model;
import org.eclipse.jetty.http.HttpStatus;
import service.AccessService;
import service.JsonService;
import spark.Request;
import spark.Response;

import java.sql.SQLException;

abstract class ModelController {
    static String index(Class<?> clazz) throws SQLException {
        return JsonService.toJson(Database.index(clazz));
    }
    static String show(Class<?> clazz, Request request) throws SQLException {
        int id;
        try {
            id = Integer.parseInt(request.params("id"));
        } catch (NumberFormatException exception) {
            return null;
        }
        return JsonService.toJson(Database.show(clazz, id));
    }

    static <T> String save(Class<T> clazz, Request request, Response response) {
        if (!AccessService.isLoggedIn(request)) {
            response.status(HttpStatus.UNAUTHORIZED_401);
            return "";
        }

        T object = JsonService.fromJson(request.body(), clazz);
        if (object == null || !Database.save(clazz, object)) {
            response.status(HttpStatus.BAD_REQUEST_400);
            return "";
        }
        return JsonService.toJson(object);
    }

    static <T extends Model> String update(Class<T> clazz, Request request, Response response) {
        if (!AccessService.isLoggedIn(request)) {
            response.status(HttpStatus.UNAUTHORIZED_401);
            return "";
        }

        int id;
        try {
            id = Integer.parseInt(request.params("id"));
        } catch (NumberFormatException exception) {
            return null;
        }

        T object = JsonService.fromJson(request.body(), clazz);
        if (object == null) {
            response.status(HttpStatus.BAD_REQUEST_400);
            return "";
        }
        object.id = id;

        if (!Database.update(clazz, object)) {
            response.status(HttpStatus.BAD_REQUEST_400);
            return "";
        }
        return JsonService.toJson(object);
    }
}

File: src/main/kotlin/controller/ModelController.kt
package controller

import exceptions.HttpBadRequest
import exceptions.HttpUnauthorized
import model.Model
import org.eclipse.jetty.http.HttpStatus
import service.AccessService
import service.JsonService
import spark.Request
import spark.Response

abstract class ModelController {
    companion object {

        @JvmStatic
        protected fun getId(request: Request): Int {
            try {
                return Integer.parseInt(request.params("id"))
            } catch (_: NumberFormatException) {
                throw HttpBadRequest("")
            }
        }

        @Suppress("UNUSED_PARAMETER")
        fun <T: Model> save(clazz: Class<T>, request: Request, response: Response): T {
            if (!AccessService.isLoggedIn(request)) {
                throw HttpUnauthorized()
            }

            val obj: T = JsonService.fromJson(request.body(), clazz) ?: throw HttpBadRequest("")
            if (obj.id != null) {
                throw HttpBadRequest("Id found when not expected")
            }
            obj.save()
            return obj
        }

        @Suppress("UNUSED_PARAMETER")
        fun <T : Model> update(clazz: Class<T>, request: Request, response: Response): T {
            if (!AccessService.isLoggedIn(request)) {
                throw HttpUnauthorized()
            }

            val obj: T = JsonService.fromJson(request.body(), clazz) ?: throw HttpBadRequest("")
            obj.id = getId(request)
            obj.save()
            return obj
        }

        fun delete(request: Request, response: Response, deleteFn: () -> Unit): String {
            if (!AccessService.isLoggedIn(request)) {
                throw HttpUnauthorized()
            }

            deleteFn()

            response.status(HttpStatus.NO_CONTENT_204)
            return ""
        }
    }
}
